fairlearn/reductions/_exponentiated_gradient/_lagrangian.py

```diff
@@ -172,7 +174,31 @@ def _call_oracle(self, lambda_vec):
             estimator = clone(estimator=self.estimator, safe=False)
 
         oracle_call_start_time = time()
-        estimator.fit(self.constraints.X, redY, **{self.sample_weight_name: redW})
+
+        # The line that is being replaced
+        # estimator.fit(self.constraints.X, redY, **{self.sample_weight_name: redW})
+
+        # ------- change made starting from here -------- #
+
+        # historical version: try and except, downside: too expensive
+        # try:
+        #     estimator.fit(self.constraints.X, redY, **{self.sample_weight_name: redW})
+        # except TypeError:
+        #     resampled_X, resampled_Y = self.resampling_method(self.constraints.X, redY, redW)
+        #     estimator.fit(resampled_X, resampled_Y)
+
+        # proposing version: if check args
+        if estimator.fit() and 'sample_weight' in inspect.getfullargspec(estimator.fit).args:
+            estimator.fit(self.constraints.X, redY, **{self.sample_weight_name: redW})
```

Comment:
it's more pythonic to put this in a try/except block instead of checking the signature of a function. A method can also accept `sample_weight` through `**kwargs` instead of `sample_weight.

Reply:
Good point. I forgot about `**kwargs` :-/

Reply:
I actually recommended not using try/except because it's more expensive when you hit the exception, so this one's on me. I suppose we don't expect this to happen terribly often anyway, and when it happens it happens once rather than thousands of times (at which point the "more expensive" argument would count more).

Reply:
It's going to happen each time through the loop (for non-sample_weight estimators).... however, it should be a fair bet that `fit()` is vastly more expensive than catching an exception.

I too was of the school that exceptions should be truly exceptional, though.

Reply:
In python we use exception handling as a way to flow-control, which is not necessarily the case in many other programming languages, here you have a nice summary of it: https://stackoverflow.com/questions/16138232/is-it-a-good-practice-to-use-try-except-else-in-python